app/javascript/src/local_level.tsx

```diff
@@ -84,11 +112,20 @@ class Tile extends React.Component {
   }
 }
 
-class LocalLevel extends React.Component {
-  constructor(props) {
+interface LocalLevelState {
+  messages: string[];
+  errors:   string[];
+  tiles:    ITile[][];
+  player:   IPlayer;
+}
+
+class LocalLevel extends React.PureComponent<{ data: LocalLevelState }> {
+  state: LocalLevelState;
```

Comment:
Not sure we need this here if types are defined on `React.PureComponent `.

Reckon you could go with something like 
```js
interface LocalLevelState {
  messages: string[];
  errors:   string[];
  tiles:    ITile[][];
  player:   IPlayer;
}

interface LocalLevelProps {
  data:     LocalLevelState;
}

class LocalLevel extends React.PureComponent<LocalLevelProps, LocalLevelState> {
  constructor(props: LocalLevelProps) {
    super(props);

    this.state = props.data;
  }
```